src/runtime_src/core/tools/xbutil2/SubCmdReset.cpp

```diff
@@ -45,7 +46,39 @@ pretty_print_action_list(xrt_core::device_collection& deviceCollection, xrt_core
 static void
 reset_device(std::shared_ptr<xrt_core::device> dev, xrt_core::query::reset_type reset)
 {
-  dev->reset(reset);
+  if (xrt_core::device_query<xrt_core::query::rom_vbnv>(dev).find("_u30_") != std::string::npos) {
+    /*
+     * u30 reset relies on working SC and SN info. SN is read and saved
+     * when FPGA is ready. so even if there is firewall trip now, we expect
+     * to be able to get S/N again
+     * Having SN info available also implies there is a working SC
+     */
+    std::string sn;
+    try {
+      sn = xrt_core::device_query<xrt_core::query::xmc_serial_num>(dev);
+    } catch (std::exception& ex) {
+      std::cerr << "ERROR:" << ex.what() << std::endl;
+      return;
+    }
+    if (sn.empty()) {
+      std::cerr << "Reset relies on S/N, but S/N can't be read from SC" << std::endl;
+      std::cout << boost::format("Reset failed on Device[%s]\n") 
+        % xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(dev));
+      return;
+    }
+    std::cout << "Card level reset. This will reset all FPGAs on the card." << std::endl;
+  }
+  //xocl reset is done through ioctl 
+  if(reset.get_key() == xrt_core::query::reset_key::user) {
+    try {
+      dev->user_reset(XCL_USER_RESET);
+    } catch(const xrt_core::error& e) {
+      std::cerr << "ERROR: " << e.what() << std::endl;
+    }
```

Comment:
The above pattern isn't followed.  Shouldn't there be a return here?

If not, then add a comment indicating that this is a "fall through" catch.

Reply:
user_reset throws an error, but here, we are catching and "cleanly" printing out the error. I'm missing a return in the catch which I have added.